docs/proposals/20210305-sidecarset-hotupgrade.md

```diff
@@ -0,0 +1,276 @@
+2. {sidecarContainer.name}-2: as shown in the figure below: envoy-2, the container run the hot upgrade empty container, and it doesn't have to deal with any real logic, as long as it stays in place, such as empty:1.0
+
+![sidecarset hotupgrade_injection](../../docs/img/sidecarset_hotupgrade_injection.png)
+
+### Hot Upgrade
+
+The SidecarSet Controller breaks down the hot upgrade process of the sidecar container into three steps:
+1. Upgrade: upgrade the empty container to the new version of the sidecar container, such as envoy-2.Image = envoy:1.17.0
+2. Migration: the process completes the state migration of stateful container, which needs to be done by the business itself according to its own container characteristics.
+PostStartHook completes the migration of the above process.
+(**Note: PostStartHook must block during the migration, and exit when migration complete.**)
+3. Reset: the step resets the old version sidecar container into empty container, such as envoy-1.Image = empty:1.0
+
+The above is the complete hot upgrade process. If a Pod needs to be hot upgraded several times, the above three steps can be repeated.
+
+![sidecarset hotupgrade](../../docs/img/sidecarset_hotupgrade.png)
+
+#### Migration Mechanism
+
+SidecarSet Hot upgrade provides an upgrade mechanism for stateful container. The specific state transfer process still needs to be implemented by the business itself.
+State migration is a very important step in the hot upgrade process, so i will describe some of the key points below:
+
+**1. Consultation Mechanism**
+
+The hot migration process starts with a negotiation between the two sidecar containers to decide the migration direction (sidecar-1 -> sidecar-2 or sidecar-2 -> sidecar-1).
+Since state migration is always from old -> new version. For this case OpenKruise injects the environment variable **SIDECARSET_VERSION** into the sidecar container by default.
+The value of SIDECARSET_VERSION is the resourceVersion of the corresponding SidecarSet, so the larger one is the newer version compared with the two SIDECARSET_VERSION. As shown below, envoy-2 is the new version of sidecar:
+```
+apiVersion: v1
+kind: Pod
+spec:
+  containers:
+  - name: envoy-1
+    env:
+    - name: SIDECARSET_VERSION
+      value: '1097936677'
+  - name: envoy-2
+    env:
+    - name: SIDECARSET_VERSION
+      value: '1098297016'
+```
+Although the version of sidecar has been saved in the environment SIDECARSET_VERSION, it still needs to be negotiated between two sidecars to determine, and recommend to use shared Linux IPC, such as: TCP communication, shared memory, UDS, etc.
+The specific business can decide by itself according to its own technology stack.
+
+**NOTE: The next release will include environment SIDECARSET_VERSION_ALT( the other sidecar version ) in the container, so negotiation is no longer required**
```

Comment:
why not give the complete design in one shot , and give a implementation plan in the proposoal?